examples/medplum-scheduling-demo/src/bots/core/block-availability.ts

```diff
@@ -0,0 +1,90 @@
+import { MedplumClient, BotEvent } from '@medplum/core';
+import { Appointment, Bundle, BundleEntry, Reference, Schedule, Slot } from '@medplum/fhirtypes';
+
+export interface BlockAvailabilityEvent {
+  schedule: Reference<Schedule>;
+  start: string;
+  end: string;
+}
+
+export async function handler(medplum: MedplumClient, event: BotEvent<BlockAvailabilityEvent>): Promise<Bundle> {
+  const { schedule, start, end } = event.input;
+
+  // Basic data validation
+  if (new Date(end) < new Date(start)) {
+    throw new Error('End date must be after start date');
+  }
+
+  const entries: BundleEntry[] = [];
+
+  // Create a Slot with status 'busy-unavailable' for the period
+  const blockedSlot: Slot = {
+    resourceType: 'Slot',
+    schedule: schedule,
+    start: start,
+    end: end,
+    status: 'busy-unavailable',
+  };
+  entries.push({
+    request: {
+      method: 'POST',
+      url: 'Slot',
+    },
+    resource: blockedSlot,
+  });
+
+  // Cancel booked appointments that overlap the period
+  const appointmentsToCancel: Appointment[] = await medplum.searchResources(
+    'Appointment',
+    `date=lt${end}&date=ge${start}&status=booked`
+  );
+  entries.push(
+    ...(appointmentsToCancel.map((appointment) => ({
+      fullUrl: `urn:uuid:${appointment.id}`,
+      request: {
+        method: 'PUT',
+        url: `Appointment?_id=${appointment.id}`,
+      },
+      resource: {
+        ...appointment,
+        status: 'cancelled',
+        cancelationReason: {
+          coding: [
+            {
+              system: 'http://terminology.hl7.org/CodeSystem/appointment-cancellation-reason',
+              code: 'prov',
+              display: 'Provider',
+            },
+          ],
+        },
+      },
+    })) as BundleEntry[])
+  );
```

Comment:
Since we're not doing an upsert (since we know this resource exists), we can instead do a pure update (PUT without a search)